src/boot_O2/loadfragment.c

```diff
@@ -1,8 +1,171 @@
 #include "global.h"
 #include "system_malloc.h"
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/loadfragment/Load_Relocate.s")
+s32 gLoadLogSeverity = 2;
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/loadfragment/Load_LoadOverlay.s")
+void Load_Relocate(void* allocatedVRamAddr, OverlayRelocationSection* ovl, u32 vRamStart) {
+    u32 sections[4];
+    u32* relocDataP;
+    u32 reloc;
+    uintptr_t relocatedAddress;
+    u32 i;
+    u32* luiInstRef;
+    uintptr_t allocu32 = (uintptr_t)allocatedVRamAddr;
+    u32* regValP;
+    u32* luiRefs[32];
+    u32 luiVals[32];
+    u32 isLoNeg;
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/loadfragment/Load_AllocateAndLoad.s")
+    if (gLoadLogSeverity >= 3) {}
+
+    sections[0] = 0;
+    sections[1] = allocu32;
+    sections[2] = allocu32 + ovl->textSize;
+    sections[3] = sections[2] + ovl->dataSize;
+
+    for (i = 0; i < ovl->nRelocations; i++) {
+        reloc = ovl->relocations[i];
+        relocDataP = (u32*)(sections[reloc >> 0x1E] + (reloc & 0xFFFFFF));
+
+        switch (reloc & 0x3F000000) {
+            case 0x2000000:
+                /* R_MIPS_32
+                 * Handles 32-bit address relocation.  Used in things such as
+                 * jump tables.
+                 */
+                if ((*relocDataP & 0xF000000) == 0) {
+                    *relocDataP = (*relocDataP - vRamStart) + allocu32;
+                } else {
+                    if (gLoadLogSeverity >= 3) {}
+                }
+                break;
+            case 0x4000000:
+                /* R_MIPS_26
+                 * Handles 26-bit address relocation, used for jumps and jals
+                 */
+                *relocDataP =
+                    (*relocDataP & 0xFC000000) |
+                    (((allocu32 + ((((*relocDataP & 0x3FFFFFF) << 2) | 0x80000000) - vRamStart)) & 0xFFFFFFF) >> 2);
+                break;
+            case 0x5000000:
+                /* R_MIPS_HI16
+                 * Handles relocation for a lui instruciton, store the reference to
+                 * the instruction, and will update it in the R_MIPS_LO16 section.
+                 */
+                luiRefs[(*relocDataP >> 0x10) & 0x1F] = relocDataP;
+                luiVals[(*relocDataP >> 0x10) & 0x1F] = *relocDataP;
+                break;
+            case 0x6000000:
+                /* R_MIPS_LO16
+                 * Updates the LUI instruction to reflect the relocated address.
+                 * The full address is calculated from the LUI and lo parts, and then updated.
+                 * if the lo part is negative, add 1 to the lui.
+                 */
```

Comment:
I mean, it also has to update the other half of the address usually 🤔 That's what the main part is doing, I suppose.

Reply:
Yeah kind of mentioned in `R_MIPS_HI16`, though I updated the comments to hopefully be more clear.